### PR TITLE
Speed up multipart callback dispatch

### PR DESCRIPTION
**Pull request by Kludex**
## Summary

- normalize omitted and explicit-`None` multipart callbacks to shared no-op functions once during parser initialization
- cache callbacks in private dispatch slots and keep them synchronized through `set_callback()`
- dispatch callbacks directly instead of constructing callback names and calling `BaseParser.callback()` for every event
- pass resolved data callbacks into the existing mark/lookbehind helper

## Performance

CodSpeed reports a **24.15% efficiency improvement** for the corrected 100-field complete-parse benchmark, from 3.7 ms to 3.0 ms. The other four corrected benchmarks are unchanged. CodSpeed notes that the compared runs used different runtime environments, so the same-environment local results below provide additional context.

Same-environment local wall-time benchmarks on Python 3.13:

| Benchmark | Base | Head | Improvement |
| --- | ---: | ---: | ---: |
| Simple multipart form | 15.53 µs | 13.63 µs | 12.2% |
| 100-field multipart form | 403.28 µs | 334.28 µs | 17.1% |
| 1,000-field multipart form | 3,988.07 µs | 3,341.27 µs | 16.2% |
| 8 MiB file upload | 605.73 µs | 599.75 µs | effectively unchanged |
| 1 MiB CR/LF-dense upload | 57.96 µs | 55.52 µs | 4.2% |

A Starlette-shaped 100-field multipart parse improved from approximately 661 µs to 616 µs (about 6.8%).

## Validation

- 162 tests pass
- 100% statement coverage
- Ruff passes
- mypy passes
- source distribution check passes

<a href="https://cubic.dev/pr/Kludex/python-multipart/pull/317?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by Kludex** [comment]
After rebasing onto the corrected benchmark lifecycle from #318, CodSpeed now reports a 24.15% efficiency improvement for the 100-field complete-parse benchmark (3.7 ms to 3.0 ms), with the other four corrected benchmarks unchanged. The report warns that the runs used different environments, while same-environment local measurements show a 17.1% improvement for the equivalent workload. The Cubic finding is marked addressed in d74fe57 and its thread is resolved; no further code change is needed.

File: python_multipart/multipart.py [hunk 1]
```diff
@@ -1119,6 +1119,35 @@ def _internal_write(self, data: bytes, length: int) -> int:
         current_header_count = self._current_header_count
         current_header_size = self._current_header_size
 
+        callbacks = cast("MultipartCallbacks", self.callbacks)
+        on_part_begin = callbacks.get("on_part_begin")
+        on_part_data = callbacks.get("on_part_data")
+        on_part_end = callbacks.get("on_part_end")
+        on_header_begin = callbacks.get("on_header_begin")
+        on_header_field = callbacks.get("on_header_field")
```
[the remark this answers is not in the document]
Comment (Kludex) [reply python_multipart/multipart.py]: Fixed in c83a120. Multipart callbacks are now cached in private dispatch slots at initialization, and `MultipartParser.set_callback()` updates both the public callback dictionary and the corresponding slot immediately. Each event reads the slot at dispatch time, so installing, replacing, or removing a callback from another callback takes effect later in the same `write()`, while avoiding per-chunk normalization overhead.

**Comment by Kludex** [comment]
The current CodSpeed comparison is not representative because the benchmark fixture reuses an already-finished parser across measured iterations. I opened #318 to create and finalize a fresh parser per benchmark invocation. PR #317 should be rebased and evaluated again after #318 merges; the corrected baseline will provide a meaningful performance comparison.

File: tests/test_multipart.py [hunk 2]
```diff
@@ -1592,6 +1592,45 @@ def test_invalid_max_size_multipart(self) -> None:
...
+            nonlocal header_begins
+            header_begins += 1
+
+        def on_header_field(data: bytes, start: int, end: int) -> None:
+            header_fields.append(data[start:end])
+
+        parser = MultipartParser(b"boundary")
+        parser.set_callback("header_begin", on_header_begin)
+        parser.set_callback("header_field", on_header_field)
+        parser.set_callback("field_start", None)
```
[the remark this answers is not in the document]
Comment (Kludex) [reply tests/test_multipart.py]: Good catch. Applied the suggested real multipart callback in d74fe57: the test now calls `set_callback("part_data", None)`. I also removed the unnecessary invalid-name branch from the private dispatch-slot synchronization. All 162 tests pass with 100% coverage.

**Comment by Kludex** [comment]
Rebased onto main after #318 merged. CodSpeed will now compare PR #317 against fresh-parser, complete-request benchmark iterations. Also addressed the Cubic review in d74fe57 by testing `set_callback("part_data", None)` and resolving the thread.
